nova_vm/src/ecmascript/types/language/string/data.rs

```diff
@@ -2,13 +2,40 @@ use wtf8::{Wtf8, Wtf8Buf};
 
 use crate::heap::{CompactionLists, HeapMarkAndSweep, WorkQueues};
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone)]
 pub struct StringHeapData {
-    // pub utf16_len: Option<u32>,
     pub(crate) data: StringBuffer,
+    pub(crate) mapping: IndexMapping,
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+impl PartialEq for StringHeapData {
+    fn eq(&self, other: &Self) -> bool {
+        // If both strings are static, we can compare their pointers directly.
+        if let (&StringBuffer::Static(self_str), &StringBuffer::Static(other_str)) =
+            (&self.data, &other.data)
+        {
+            if std::ptr::eq(self_str, other_str) {
+                return true;
+            }
+        }
+        self.as_str() == other.as_str()
+    }
+}
+impl Eq for StringHeapData {}
+
+#[derive(Debug, Clone)]
+pub(crate) enum IndexMapping {
+    Ascii,
+    NonAscii {
+        /// Mapping from UTF-16 indices to indices in the UTF-8 representation.
+        /// When the UTF-16 character would be the second character in a
+        /// surrogate pair, it maps to None because there is no corresponding
+        /// UTF-8 index.
+        mapping: Box<[Option<usize>]>,
```

Comment:
**thought:** I think we'll eventually want to restrict the UTF-8 length of strings to 2^32 even though that means that some 2^32 UTF-16 strings become too long and thus not spec-compatible. Then this would become `Option<u32>`.

Actually: This could maybe already be `Option<NonZeroUsize>`; Only zero will map to zero anyhow so we don't really need to keep that around.

Reply:
The spec's maximum string length is not 2^32, but 2^53 - 1. Restricting the length further in an engine seems fine –after all, clearly you could not support strings longer than 2^32 in a 32-bit system– but is that needed? With `usize`, in 64-bit systems you would still have room to represent the maximum spec string in UTF-8.

As for `Option<NonZeroUsize>`, I do like the fact that currently `mapping[u16_idx]` will just turn it into a UTF-8 index with no extra work. But that's not worth the extra memory size. Were you thinking of having `mapping[0]` always map to `None`, or of offsetting the indices by one?

Reply:
I was thinking... Something abstract probably.

But in basic terms the idea was that we'd never ask `mapping[0]` because that's trivially always 0, and then `mapping[1]` would map to either `None` or `Some(NonZeroUsize)`: It shouldn't be possible for some non-zero index to map to zero after all.

Reply:
`mapping[0] = None` does seem like a solid idea now that I finally understood the issue with `mapping[0]` :)